File: pmvs10b-lab6-grimb000/NumberGuess/app/src/main/java/com/zizto/numberguess/MainActivity.java
package com.zizto.numberguess;

import android.content.SharedPreferences;
import android.gesture.Gesture;
import android.gesture.GestureLibraries;
import android.gesture.GestureLibrary;
import android.gesture.GestureOverlayView;
import android.gesture.Prediction;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.view.View;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import java.util.ArrayList;

public class MainActivity extends AppCompatActivity implements GestureOverlayView.OnGesturePerformedListener {

    Button bControl;
    EditText etInput;
    TextView tvInfo, tvAttempts;
    ImageView ivIcon;
    ProgressBar progressBar;

    GestureLibrary gLib;
    GestureOverlayView gesturesView;

    int numberToGuess = 0;
    boolean isGameEnded = false;
    int attemptsCount = 0;
    final int MAX_ATTEMPTS = 10;
    int launchCount = 0;

    SharedPreferences appSettings;
    String uniqueAppId;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        appSettings = getSharedPreferences("app_settings", MODE_PRIVATE);
        SharedPreferences.Editor editor = appSettings.edit();

        launchCount = appSettings.getInt("launch_count", 0) + 1;
        editor.putInt("launch_count", launchCount);

        uniqueAppId = appSettings.getString("app_id", null);
        if (uniqueAppId == null) {
            uniqueAppId = java.util.UUID.randomUUID().toString();
            editor.putString("app_id", uniqueAppId);
        }

        editor.apply();

        tvInfo = findViewById(R.id.textView);
        tvAttempts = findViewById(R.id.tvAttempts);
        etInput = findViewById(R.id.editTextNumber);
        bControl = findViewById(R.id.button);
        ivIcon = findViewById(R.id.ivIcon);
        progressBar = findViewById(R.id.progressBarAttempts);

        gLib = GestureLibraries.fromRawResource(this, R.raw.gestures);
        if (!gLib.load()) {
            Toast.makeText(this, "Не удалось загрузить жесты!", Toast.LENGTH_SHORT).show();
            finish();
        }

        gesturesView = findViewById(R.id.gestures_overlay);
        gesturesView.addOnGesturePerformedListener(this);

        startNewGame();
    }

    @Override
    public void onGesturePerformed(GestureOverlayView overlay, Gesture gesture) {
        ArrayList<Prediction> predictions = gLib.recognize(gesture);

        if (predictions.size() > 0 && predictions.get(0).score > 1.0) {
            String gestureName = predictions.get(0).name;

            if (isNumeric(gestureName)) {
                etInput.append(gestureName);
            }
            else if (gestureName.equals("check")) {
                onClick(bControl);
            }
            else if (gestureName.equals("clear")) {
                etInput.setText("");
            }
            else {
                Toast.makeText(this, "Жест: " + gestureName, Toast.LENGTH_SHORT).show();
            }
        }
    }

    private boolean isNumeric(String str) {
        try {
            Integer.parseInt(str);
            return true;
        } catch (NumberFormatException e) {
            return false;
        }
    }

    private void startNewGame() {
        numberToGuess = (int) (Math.random() * 100);
        isGameEnded = false;
        attemptsCount = 0;

        bControl.setText(getResources().getString(R.string.input_value));
        tvInfo.setText(getResources().getString(R.string.try_to_guess));
        tvInfo.setText("Id = " + uniqueAppId + "; kolZap = " + launchCount);
        etInput.setText("");

        progressBar.setProgress(0);
        String attemptsText = getResources().getQuantityString(
                R.plurals.attempts_count,
                attemptsCount,
                attemptsCount,
                MAX_ATTEMPTS
        );
        tvAttempts.setText(attemptsText);

        bControl.setEnabled(true);

        Animation combo = AnimationUtils.loadAnimation(this, R.anim.combo_anim);
        ivIcon.startAnimation(combo);
    }

    public void onClick(View v) {
        if (isGameEnded) {
            startNewGame();
            return;
        }

        String inputText = etInput.getText().toString();
        if (inputText.isEmpty()) {
            tvInfo.setText(getResources().getString(R.string.error_empty));
            return;
        }

        int inp = Integer.parseInt(inputText);
        if (inp < 0 || inp > 100) {
            tvInfo.setText(getResources().getString(R.string.error_range));
            etInput.setText("");
            return;
        }

        Animation rotate = AnimationUtils.loadAnimation(this, R.anim.rotate_center);
        ivIcon.startAnimation(rotate);

        bControl.setEnabled(false);

        new Handler(Looper.getMainLooper()).postDelayed(() -> {
            checkGuess(inp);
            if (!isGameEnded) {
                bControl.setEnabled(true);
            }
        }, 600);
    }

    private void checkGuess(int inp) {
        attemptsCount++;
        progressBar.setProgress(attemptsCount);

        String attemptsText = getResources().getQuantityString(
                R.plurals.attempts_count,
                attemptsCount,
                attemptsCount,
                MAX_ATTEMPTS
        );
        tvAttempts.setText(attemptsText);

        Animation fadeIn = AnimationUtils.loadAnimation(this, R.anim.fade_in);
        tvInfo.startAnimation(fadeIn);

        if (inp == numberToGuess) {
            tvInfo.setText(getResources().getString(R.string.hit));

            int bestScore = appSettings.getInt("best_score", Integer.MAX_VALUE);
            if (attemptsCount < bestScore) {
                appSettings.edit().putInt("best_score", attemptsCount).apply();
                Toast.makeText(this, "Новый рекорд: " + attemptsCount + " попыток!", Toast.LENGTH_LONG).show();
            }

            showResultDialog("Победа!", "Вы угадали число " + numberToGuess + " за " + attemptsCount + " попыток.");
            isGameEnded = true;
            bControl.setText(getResources().getString(R.string.play_more));
            bControl.setEnabled(true);
        } else {
            if (attemptsCount >= MAX_ATTEMPTS) {
                showResultDialog("Проигрыш", "Попытки закончились! Число было: " + numberToGuess);
                isGameEnded = true;
                bControl.setText(getResources().getString(R.string.play_more));
                bControl.setEnabled(true);
            } else {
                if (inp > numberToGuess)
                    tvInfo.setText(getResources().getString(R.string.ahead));
                else
                    tvInfo.setText(getResources().getString(R.string.behind));

                etInput.setText("");
            }
        }
    }

    private void showResultDialog(String title, String message) {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle(title)
                .setMessage(message)
                .setCancelable(false)
                .setIcon(android.R.drawable.ic_dialog_info)
                .setPositiveButton("Начать заново", (dialog, id) -> startNewGame())
                .setNegativeButton("Выход", (dialog, id) -> finish());
        AlertDialog alert = builder.create();
        alert.show();
    }
}

File: pmvs10b-lab5-grimb000/GuessNumber_Kotlin/app/src/main/java/com/example/guessnumber_kotlin/MainActivity.kt
package com.example.guessnumber_kotlin

import android.gesture.Gesture
import android.gesture.GestureLibraries
import android.gesture.GestureLibrary
import android.gesture.GestureOverlayView
import android.gesture.Prediction
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.view.animation.AnimationUtils
import android.widget.*
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import coil.load
import coil.transform.CircleCropTransformation
import kotlinx.coroutines.*

class MainActivity : AppCompatActivity(), GestureOverlayView.OnGesturePerformedListener {
    private var secret = 0
    private var attempts = 0
    private val maxAttempts = 10
    private var gestureBuffer = ""

    private lateinit var ivImage: ImageView
    private lateinit var tvInfo: TextView
    private lateinit var tvAttempts: TextView
    private lateinit var tvGestureBuffer: TextView
    private lateinit var etInput: EditText
    private lateinit var bCheck: Button
    private lateinit var bRestart: Button
    private lateinit var pbAttempts: ProgressBar
    private lateinit var pbLoading: ProgressBar
    private lateinit var seekBar: SeekBar
    private lateinit var tvSeekBarValue: TextView
    private lateinit var gestureOverlay: GestureOverlayView
    private lateinit var gestureStatus: TextView
    private lateinit var gestureLibrary: GestureLibrary

    private val imageUrl = "https://img.freepik.com/free-vector/guess-number-concept-illustration_114360-1234.jpg"

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        setContentView(R.layout.activity_main)
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }

        initViews()
        setupListeners()
        setupGestures()
        startNewGame()
        loadRemoteImage()
    }

    private fun initViews() {
        ivImage = findViewById(R.id.imageView)
        tvInfo = findViewById(R.id.textViewInfo)
        tvAttempts = findViewById(R.id.textViewAttempts)
        tvGestureBuffer = findViewById(R.id.textViewGestureBuffer)
        etInput = findViewById(R.id.editTextGuess)
        bCheck = findViewById(R.id.buttonCheck)
        bRestart = findViewById(R.id.buttonRestartMain)
        pbAttempts = findViewById(R.id.progressBarAttempts)
        pbLoading = findViewById(R.id.loadingSpinner)
        seekBar = findViewById(R.id.seekBarHelper)
        tvSeekBarValue = findViewById(R.id.textViewSeekBarValue)
        gestureOverlay = findViewById(R.id.gestureOverlay)
        gestureStatus = findViewById(R.id.textViewGestureStatus)

        pbAttempts.max = maxAttempts
    }

    private fun setupListeners() {
        bCheck.setOnClickListener { checkGuess() }
        bRestart.setOnClickListener { startNewGame() }
        etInput.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) = Unit

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                syncGestureBufferFromInput(s?.toString().orEmpty())
            }

            override fun afterTextChanged(s: Editable?) = Unit
        })

        seekBar.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar?, progress: Int, fromUser: Boolean) {
                tvSeekBarValue.text = progress.toString()
                if (fromUser) etInput.setText(progress.toString())
            }
            override fun onStartTrackingTouch(seekBar: SeekBar?) {}
            override fun onStopTrackingTouch(seekBar: SeekBar?) {}
        })
    }

    private fun setupGestures() {
        gestureLibrary = GestureLibraries.fromRawResource(this, R.raw.gestures)
        if (gestureLibrary.load()) {
            gestureOverlay.addOnGesturePerformedListener(this)
            gestureStatus.text = "Draw digits or use stop to submit."
        } else {
            gestureStatus.text = "Gesture library failed to load."
        }
    }

    private fun startNewGame() {
        secret = (1..100).random()
        attempts = 0
        gestureBuffer = ""
        updateAttemptsUI()
        tvInfo.text = getString(R.string.try_to_guess)
        etInput.text.clear()
        updateGestureBuffer()
        bCheck.isEnabled = true
        bRestart.visibility = View.GONE
        
        MainScope().launch {
            pbAttempts.progress = 0
            for (i in 0..maxAttempts) {
                delay(20)
                pbAttempts.progress = i
            }
        }
        
        applyAnimation(ivImage, R.anim.entrance)
    }

    private fun loadRemoteImage() {
        pbLoading.visibility = View.VISIBLE
        ivImage.load(imageUrl) {
            crossfade(true)
            transformations(CircleCropTransformation())
            listener(
                onSuccess = { _, _ -> pbLoading.visibility = View.GONE },
                onError = { _, _ -> pbLoading.visibility = View.GONE }
            )
        }
    }

    private fun checkGuess() {
        val input = etInput.text.toString().toIntOrNull()
        if (input == null || input !in 1..100) {
            tvInfo.text = getString(R.string.error)
            applyAnimation(etInput, R.anim.shake)
            return
        }

        attempts++
        updateAttemptsUI()

        when {
            input < secret -> {
                tvInfo.text = getString(R.string.behind)
                applyAnimation(tvInfo, R.anim.hint_bounce)
            }
            input > secret -> {
                tvInfo.text = getString(R.string.ahead)
                applyAnimation(tvInfo, R.anim.hint_bounce)
            }
            else -> {
                showEndDialog(true)
            }
        }

        if (attempts >= maxAttempts && input != secret) {
            showEndDialog(false)
        }

        gestureBuffer = ""
        updateGestureBuffer()
    }

    private fun updateAttemptsUI() {
        val left = maxAttempts - attempts
        tvAttempts.text = getString(R.string.attempts_left, left)
        pbAttempts.progress = left
    }

    private fun showEndDialog(isWin: Boolean) {
        bCheck.isEnabled = false
        bRestart.visibility = View.VISIBLE

        val message = if (isWin) {
            getString(R.string.win_message, secret, attempts)
        } else {
            getString(R.string.lose_message, secret)
        }

        AlertDialog.Builder(this)
            .setTitle(R.string.game_over)
            .setMessage(message)
            .setPositiveButton(R.string.ok) { dialog, _ -> dialog.dismiss() }
            .setCancelable(false)
            .show()
            
        if (isWin) applyAnimation(ivImage, R.anim.win_celebration)
    }

    private fun applyAnimation(view: View, animResId: Int) {
        val animation = AnimationUtils.loadAnimation(this, animResId)
        view.startAnimation(animation)
    }

    override fun onGesturePerformed(overlay: GestureOverlayView?, gesture: Gesture?) {
        if (gesture == null) return

        val bestPrediction = gestureLibrary
            .recognize(gesture)
            .maxByOrNull(Prediction::score)

        if (bestPrediction == null || bestPrediction.score <= 1.0) {
            gestureStatus.text = "Unknown gesture"
            return
        }

        handleGesture(bestPrediction.name.trim())
    }

    private fun handleGesture(name: String) {
        when {
            name.length == 1 && name[0].isDigit() -> appendGestureDigit(name)
            name.equals("stop", ignoreCase = true) || name == "=" -> {
                gestureStatus.text = "Submitting gesture input"
                checkGuess()
            }
            name.equals("c", ignoreCase = true) -> {
                gestureBuffer = ""
                etInput.text.clear()
                updateGestureBuffer()
                gestureStatus.text = "Gesture input cleared"
            }
            else -> gestureStatus.text = "Gesture '$name' is not used here"
        }
    }

    private fun appendGestureDigit(digit: String) {
        if (gestureBuffer.length >= 3) {
            gestureStatus.text = "Too many digits"
            return
        }

        gestureBuffer += digit
        etInput.setText(gestureBuffer)
        etInput.setSelection(etInput.text.length)
        updateGestureBuffer()
        gestureStatus.text = "Recognized gesture: $digit"
    }

    private fun updateGestureBuffer() {
        val shownBuffer = if (gestureBuffer.isEmpty()) "-" else gestureBuffer
        tvGestureBuffer.text = getString(R.string.gesture_buffer_value, shownBuffer)
    }

    private fun syncGestureBufferFromInput(text: String) {
        val normalized = text.filter(Char::isDigit).take(3)
        if (gestureBuffer != normalized) {
            gestureBuffer = normalized
            updateGestureBuffer()
        }
    }
}
